File: app/src/main/java/app/safeinvoice/data/files/LocalStorage.kt
package app.safeinvoice.data.files

import android.content.Context
import app.safeinvoice.data.entity.FolderType
import java.io.File
import java.io.InputStream

class LocalStorage(context: Context) {
    private val root: File = File(context.filesDir, "businesses").apply { mkdirs() }

    fun businessDir(businessId: String): File =
        File(root, businessId).apply { mkdirs() }

    fun customerDir(businessId: String, customerId: String): File =
        File(businessDir(businessId), "customers/$customerId").apply { mkdirs() }

    fun folder(businessId: String, customerId: String, type: FolderType): File =
        File(customerDir(businessId, customerId), type.name.lowercase()).apply { mkdirs() }

    fun templatesDir(businessId: String): File =
        File(businessDir(businessId), "templates").apply { mkdirs() }

    fun logosDir(businessId: String): File =
        File(businessDir(businessId), "logos").apply { mkdirs() }

    fun relativeToFiles(file: File, context: Context): String =
        file.relativeTo(context.filesDir).path

    fun resolve(context: Context, relativePath: String): File =
        File(context.filesDir, relativePath)

    fun copyInto(input: InputStream, dest: File): File {
        dest.parentFile?.mkdirs()
        dest.outputStream().use { out -> input.copyTo(out) }
        return dest
    }

    fun uniqueFile(dir: File, baseName: String): File {
        val safe = baseName.replace(Regex("[^A-Za-z0-9._-]"), "_")
        var candidate = File(dir, safe)
        if (!candidate.exists()) return candidate
        val dot = safe.lastIndexOf('.')
        val stem = if (dot > 0) safe.substring(0, dot) else safe
        val ext = if (dot > 0) safe.substring(dot) else ""
        var i = 2
        while (candidate.exists()) {
            candidate = File(dir, "${stem}_$i$ext")
            i++
        }
        return candidate
    }
}

File: lib/data/local_storage.dart
import 'dart:io';

import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';

import '../domain/models.dart';

class LocalStorage {
  LocalStorage(this.root);

  final Directory root;

  static Future<LocalStorage> open() async {
    final support = await getApplicationSupportDirectory();
    final dir = Directory(p.join(support.path, 'businesses'));
    await dir.create(recursive: true);
    return LocalStorage(support);
  }

  Directory businessDir(String businessId) {
    final dir = Directory(p.join(root.path, 'businesses', businessId));
    dir.createSync(recursive: true);
    return dir;
  }

  Directory customerDir(String businessId, String customerId) {
    final dir = Directory(p.join(businessDir(businessId).path, 'customers', customerId));
    dir.createSync(recursive: true);
    return dir;
  }

  Directory folder(String businessId, String customerId, FolderType type) {
    final dir = Directory(p.join(customerDir(businessId, customerId).path, type.name));
    dir.createSync(recursive: true);
    return dir;
  }

  Directory templatesDir(String businessId) {
    final dir = Directory(p.join(businessDir(businessId).path, 'templates'));
    dir.createSync(recursive: true);
    return dir;
  }

  Directory logosDir(String businessId) {
    final dir = Directory(p.join(businessDir(businessId).path, 'logos'));
    dir.createSync(recursive: true);
    return dir;
  }

  String relativeToRoot(File file) => p.relative(file.path, from: root.path);

  File resolve(String relativeOrAbsolute) {
    final asIs = File(relativeOrAbsolute);
    if (asIs.isAbsolute) return asIs;
    return File(p.join(root.path, relativeOrAbsolute));
  }

  Future<File> copyBytes(List<int> bytes, File dest) async {
    await dest.parent.create(recursive: true);
    return dest.writeAsBytes(bytes, flush: true);
  }

  Future<File> copyFile(File source, File dest) async {
    await dest.parent.create(recursive: true);
    return source.copy(dest.path);
  }

  File uniqueFile(Directory dir, String baseName) {
    final safe = baseName.replaceAll(RegExp(r'[^A-Za-z0-9._-]'), '_');
    var candidate = File(p.join(dir.path, safe));
    if (!candidate.existsSync()) return candidate;
    final dot = safe.lastIndexOf('.');
    final stem = dot > 0 ? safe.substring(0, dot) : safe;
    final ext = dot > 0 ? safe.substring(dot) : '';
    var i = 2;
    while (candidate.existsSync()) {
      candidate = File(p.join(dir.path, '${stem}_$i$ext'));
      i++;
    }
    return candidate;
  }
}

String guessMime(String name) {
  final n = name.toLowerCase();
  if (n.endsWith('.pdf')) return 'application/pdf';
  if (n.endsWith('.png')) return 'image/png';
  if (n.endsWith('.jpg') || n.endsWith('.jpeg')) return 'image/jpeg';
  if (n.endsWith('.webp')) return 'image/webp';
  if (n.endsWith('.xlsx')) {
    return 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet';
  }
  if (n.endsWith('.xls')) return 'application/vnd.ms-excel';
  if (n.endsWith('.csv')) return 'text/csv';
  if (n.endsWith('.txt')) return 'text/plain';
  return 'application/octet-stream';
}
